File: patternlets/MPI/07.parallelLoop-chunksOf1/textual/parallelLoopChunksOf1.c
/* parallelLoopChunksOf1.c
 * ... illustrates the parallel for loop pattern in MPI 
 *	in which processes perform the loop's iterations in 'chunks' 
 *      of size 1 (simple, and useful when loop iterations 
 *      do not access memory/cache locations) ...
 * Note this is much simpler than the 'equal chunks' loop.
 * Joel Adams, Calvin College, November 2009.
 *
 * Usage: mpirun -np N ./parallelForSlices
 *
 * Exercise:
 * - Compile and run, varying N: 1, 2, 3, 4, 5, 6, 7, 8
 * - Change REPS to 16, save, recompile, rerun, varying N again.
 * - Explain how this pattern divides the iterations of the loop
 *    among the processes.
 */

#include <stdio.h>  // printf()
#include <mpi.h>    // MPI

int main(int argc, char** argv) {
    const int REPS = 8;
    int id = -1, numProcesses = -1, i = -1;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &id);
    MPI_Comm_size(MPI_COMM_WORLD, &numProcesses);

    for (i = id; i < REPS; i += numProcesses) {
        printf("Process %d is performing iteration %d\n", id, i);
    }

    MPI_Finalize();
    return 0;
}

File: patternlets/java-OpenMPI/03.parallelLoop-chunksOf1/ParallelLoopChunksOf1.java
/* ParallelLoopChunksOf1.java
 * ... illustrates the parallel loop pattern in OpenMPI+Java,
 *      in which processes perform the loop's iterations in 'chunks'
 *      of size 1 (simple, and useful when loop iterations
 *      do not access memory/cache locations) ...
 * Note this is much simpler than the 'equal chunks' loop.
 *
 * Joel Adams, Calvin University, November 2019.
 *
 * Usage: mpirun -np 4 java ./ParallelLoopChunksOf1
 *
 * Exercise:
 * - Compile and run, varying N: 1, 2, 3, 4, 5, 6, 7, 8
 * - Change REPS to 16, save, recompile, rerun, varying N again.
 * - Explain how this pattern divides the iterations of the loop
 *    among the processes.
 */

import mpi.*;

public class ParallelLoopChunksOf1 {

 public static final int REPS = 8;
 public static final int MASTER = 0;

 public static void main(String [] args) throws MPIException {
    MPI.Init(args);

    int id           = MPI.COMM_WORLD.getRank();
    int numProcesses = MPI.COMM_WORLD.getSize();
    String message   = "";

    if (numProcesses > REPS) {
      if (id == MASTER) {
          System.out.print("\nPlease run with -np less than or equal to "
                  +  REPS + "\n\n");
      }
    } else {
      for (int i = id; i < REPS; i += numProcesses) { 
          message = "Process " + id + " is performing iteration " + i + "\n";
          System.out.print(message);
      }
    }

    MPI.Finalize();
  }
}

